src/openforms/submissions/exports.py

```diff
@@ -25,23 +31,44 @@ class ExportFileTypes:
     XML = FileType("xml", "text/xml")
 
 
+def iter_submission_data_nodes(submission: Submission) -> Iterator[Node]:
+    renderer = Renderer(submission, mode=RenderModes.export, as_html=False)
+    for step_node in renderer.get_children():
+        for component_node in step_node.get_children():
+            if component_node.is_layout:
+                continue
+            yield component_node
+
+
 def create_submission_export(queryset: SubmissionQuerySet) -> tablib.Dataset:
-    headers = []
-    for submission in queryset:
-        headers += list(submission.get_merged_data().keys())
-    headers = list(dict.fromkeys(headers))  # Remove duplicates
-    data = tablib.Dataset(headers=["Formuliernaam", "Inzendingdatum"] + headers)
+    """
+    Turn a submissions queryset into a tablib dataset for export.
+
+    .. note:: the queryset of submissions must all be of the same form!
+    """
+    # queryset *could* be empty
+    if not queryset:
+        return tablib.Dataset()
+
+    first_submission = queryset[0]
+    headers = ["Formuliernaam", "Inzendingdatum"]
+    for component_node in iter_submission_data_nodes(first_submission):
```

Comment:
I think there could be problems in the case where the definition of a form changes, and an admin tries to export multiple submissions of the same form (but that have different definitions).

Reply:
hrm, good point, let's see...

1. Form fields have been removed that are present in submissions -> that's okay, apparently they weren't relevant otherwise they wouldn't be removed
2. Form fields were added that are missing in old submissions -> should not be a problem, export contains more headers, it's as if those fields were not filled out
3. Form fields have changed component keys - in that case, the downstream systems processing it will have issues too. Not sure if we can/should support this. And maybe variables may be the solution for this?

@joeribekker thoughts?

Reply:
Discussed with Joeri - we'll see if/when someone complains about this and #1586 is created to track this.